mgr/dashboard: Task wrapper service

Devp00l opened the pull request:
Has a method to wrap an API call into a task.

These changes are also used in the upcoming Pool Management PR https://github.com/ceph/ceph/pull/21614.

Signed-off-by: Stephan Müller <smueller@suse.com>

ricardoasmarques commented:
@Devp00l  Please add a separate commit adapting the RBD implementation.

ricardoasmarques commented:
@Devp00l is this PR fixing issue https://tracker.ceph.com/issues/24134 ?

If yes, you should add `Fixes https://tracker.ceph.com/issues/24134` to the commit message.

Devp00l commented:
jenkins retest this please

Devp00l commented:
@ricardoasmarques Added the RBD implementations as separate commits :)

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-wrapper.service.ts:
```diff
@@ -0,0 +1,71 @@
+  _handleExecutingTasks(task: FinishedTask, tasks?: ExecutingTask[]) {
+    this.notificationService.show(
+      NotificationType.info,
+      task.name + ' in progress...',
```
`task.name` is an internal `identifier` that should not be displayed to the user.

This is what you currently have implemented:

![screenshot from 2018-06-01 13-38-05](https://user-images.githubusercontent.com/14297426/40841187-77e44a36-65a1-11e8-830f-fffb4a456047.png)

Intead of `rbd/create in progress....` I would recommend a different language here, like `RBD creation in progress...` (just like we had before this refactor).

You can implement this by adding a new `string` to the `TaskManagerMessage` class, to let the developer specify the "in progress" message for each task.

Devp00l commented:
Addressed your comment @ricardoasmarques 

Question should I introduce the following changes in this PR or a separate:
1. Renaming `taskManagerMessageService` to `taskMessageService`

2. Reducing messages to follow the following style:
Running ->  "Creating ..."
Success -> "Created ..."
Error ->  "Create ... failed" + error

I'm in favor of separating it.

ricardoasmarques commented:
@Devp00l since you use separate commits, I'm fine with both.

But if you are in favor of a separate PR, let's do it in a separate PR after merging this one.

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -56,81 +61,104 @@ export class TaskManagerMessageService {
-                    `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}' ` +
-                    `has been created successfully`,
+      (metadata) =>
+        `Create snapshot` +
```
Missing a "space" before snapshot name:

![screenshot_20180604_161656](https://user-images.githubusercontent.com/14297426/40925957-e4dd368e-6812-11e8-8ee6-c6344d19d17b.png)

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -56,81 +61,104 @@ export class TaskManagerMessageService {
+        `Create snapshot` +
+        `'${metadata.pool_name}/${metadata.image_name}@${metadata.snapshot_name}'`,
+      (metadata) =>
+        `Creating snapshot` +
```
Missing a "space" before snapshot name.

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -139,14 +167,18 @@ export class TaskManagerMessageService {
     },
+    (metadata) => {
+      return (
+        'Progressing ' + (Components[metadata.component] || metadata.component || 'unknown task')
```
For consistency across all task-related messages, I recommend you to use `Executing` instead of `Progressing`.

Devp00l commented:
Addressed all changes :)

Devp00l commented:
jenkins retest this please

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -56,81 +61,104 @@ export class TaskManagerMessageService {
       (metadata) => `Copy RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'`,
-      (metadata) => `RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'
-                     has been copied successfully`,
+      (metadata) => `Coping RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'`,
```
Both `title`and `running` message start with similar words:

![screenshot from 2018-06-05 17-48-03](https://user-images.githubusercontent.com/14297426/40990477-dd34a9a4-68e8-11e8-92c3-953f7a6f85d4.png)

Personally, I prefer the "language" we had before:

![screenshot from 2018-06-05 17-47-24](https://user-images.githubusercontent.com/14297426/40990504-f35f630e-68e8-11e8-98d2-ac5022cdac9a.png)

But this is just a personal preference.

ricardoasmarques replied:
Typo - s/Coping/Copying/

Devp00l replied:
Now the behavior is different - You will only see `Coping RBD 'rbd/i5'` as running task.
But you are showing the Notification tab here - as I mentioned I wanted to change the behavior in another PR, in order to have similar messages.

Devp00l replied:
Fixed typo :)

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.ts:
```diff
@@ -56,81 +61,104 @@ export class TaskManagerMessageService {
       (metadata) => `Copy RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'`,
-      (metadata) => `RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'
-                     has been copied successfully`,
+      (metadata) => `Coping RBD '${metadata.dest_pool_name}/${metadata.dest_image_name}'`,
```
Missing `...` in the end of the `running`message.

Devp00l commented:
@ricardoasmarques  I removed the three dots as three dots are used in application to set the expectation that there is more, like in a menu bar - if you see an entry with three dots you will expect a new window with settings. On blogs or news sites three dot's are often used to show the continuation of the article - and mostly something pops up if you click on them asking you to subscribe.

The message is also shown right to an loading icon. So I assumed the three dot's a more irritating than helpful.

ricardoasmarques commented on src/pybind/mgr/dashboard/frontend/src/app/core/navigation/task-manager/task-manager.component.ts:
```diff
@@ -11,38 +11,44 @@ import { TaskManagerMessageService } from '../../../shared/services/task-manager
+
+  _setIcon(executingTasks: number) {
+    const iconSuffix = ['o', 'start', 'half', 'end'];
+    // TODO: Use all suffixes
```
Shouldn't we remove this comment?

Devp00l replied:
We are currently not using all suffixes - I could remove the unused ones, but they were implemented first - so I think someone wanted to use them.

Devp00l replied:
Removed all other comments except the TODO

Devp00l commented:
jenkins retest this please

Devp00l commented:
@ricardoasmarques removed all comments expect the TODO

Devp00l commented:
jenkins retest this please

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-wrapper.service.ts:
```diff
@@ -0,0 +1,74 @@
+      executingTask.name,
+      executingTask.metadata,
+      (asyncTask: FinishedTask) => {
+        this.notificationService.notifyTask(asyncTask);
```
Could you try to cover this line in unit tests?

tspmelo commented on src/pybind/mgr/dashboard/frontend/src/app/shared/services/task-manager-message.service.spec.ts:
```diff
@@ -28,6 +28,25 @@ describe('TaskManagerMessageService', () => {
+
   it('should getErrorMessage', () => {
     finishedTask.exception = _.assign(new TaskException(), {
       code: 1
```
could you add the following in line 68: `expect(value.running({})).toBeTruthy();`

Devp00l commented:
@tspmelo I've addressed all your requested changes
